### DISCUSSION TITLE
api rule to allow based on field value?

### DISCUSSION THREAD

**Discussion by ccbiozhaw** [pocketbase/pocketbase#3809] [Q&A]
hey. i have a collection called events which essentially contains the following information 
```
event {
// random stuff
device: someDevice 
}
```

i want to create a second collecton now, which is essentially:

```
devices {
someDevice: userIds[]
}
```

how can i create an api rule that checks for the type of device in the device column, then check if that user is in the devices collection to create the rule?

essentially i want restrict which users can add an event depending on the device type.
thanks

**Comment by ganigeorgiev** [comment]
I'm sorry, I don't understand what is the question, nor what is the desired api rule.

Please try to elaborate in more details on:
- what exactly is your collections structure (you can also export your current collections configuration as JSON from the _Admin UI > Settings > Export collections_)
- what is the desired constraint (you mentioned "_event depending on the device type_" but there is no device type in the above and what it has to do with the userIds?)
- for what rules (list, view create, update, delete) and for which collection you wan to apply them

**Reply by ccbiozhaw** [reply]
ok I have a collection i called events. Events contain a bunch of information, including one field called devices. devices are just a string, which can be dynamically generated so it could be various.
what i now want is to manually allow only certain users to add new events for certain devices. for example USER A might only be able to work on deviceA, and device B, while USER B can edit all devices and so forth.


i thought what i need to do is make a new collection, lets call it permissions where i specify the devices that exist and a relation to the users that can edit them. So in the example above:

collection: Permissions

```
{
deviceA: [user A],
deviceB: [user A, user B],
deviceC:[user B],
}
```


essentially the api rould would be something like @collection.permissions.(device from field of record to be added) includes the id of @request.author.id ?? or something like that

**Reply by ganigeorgiev** [reply]
This is really confusing and doesn't really help understanding your data structure, nor what you are trying to do.

The `deviceA deviceB ...` code example doesn't really make sense to me.

> devices are just a string, which can be dynamically generated so it could be various.

But how this devices will be stored and in what field type? Would this be a separate collection?
How they will be related to the users?
What is even `permissions` and what fields it has?

Also note that if you use a `text` and not `relation` field to store the device in the events, then relations in PocketBase will not work because we support only relations via a record `id`.

If you need further help, please try to provide more clear description of your use case and elaborate a little more on your actual data structure. I just don't feel confident providing suggestions for security related topics if there is no clear understanding of the desired behavior.

**Reply by ccbiozhaw** [reply]
ok i work on a simple booking application for an internal problem. Essentially we have different devices that people want to book. So i have one collection called events where i store information abuot the booking, as well as the device that was booked. For example an entry could be:

```
device: freezer
start: today
end: tomorrow
```

an other entry could be 

```
device: centrifuge
start: today
end: tomorrow
```

and so forth. 


now, i do not want every user to book on every device. for example, we might collaborate with other people, and i want them to be able to access some devices but not others. 

does this help?

**Reply by ccbiozhaw** [reply]
i think i am close with the following rule: 

@collection.groups.freezer.id ?= @request.auth.id

however, instead of freezer, i want to have the device type of the record i intend to create

**Reply by ganigeorgiev** [reply]
Again, it is not clear from your descriptions what collections and what fields do you have and what is the relations between them.

**The API rules depends on how your collections are defined and the above can be expressed in several different ways depending on the desired access patterns.**

You refuse to elaborate on the exact data structure you have so I'll try as a last attempt to propose a structure based on what I understand from your comments: 

| users |
| --- |
| id |
| ... |

| events |
| --- |
| id |
| device (single `relation` to "devices") |
| ... |

| devices |
| --- |
| id |
| name |
| start (`date`) |
| end (`date`) |
| users (multiple `relation` to users) |
| ... |

(_the above "users" relation eventually can be also expressed as a separate `deviceUser` collection with 2 single relation fields - one to the users and another to the devices or even a `devices` multiple `relation` field in the `users` collection, but this all will depend on how you want to interact with your data_)

And now for example if we want to define a create API rule for the `events` collection in order to restrict
users from creating events to devices that they don't have relation to, you can use something like:

```js
@request.auth.id != "" &&
device.users.id ?= @request.auth.id
```

But again I don't fully understand the exact use case and its requirements.

I'll stop pursuing the matter further as it seems that my questions are just causing more confusion, but if you want others to understand better what you are trying to do, please specify the exact collections structure you currently have and what is the desired outcome (pseudo code, erd diagrams, or even example data flow can also help).

**Reply by ccbiozhaw** [reply]
sorry for the confusion:s thank you very much for your help though
